### PR TITLE
Pagination: automatically remove sibling links as space permits

### LINKED ISSUE #1239: Enhance Pagination to dynamically adjust sibling links
**Enhancement** 

Enhance the Pagination widget to automatically remove sibling links when there isn't space to include them. `siblingCount` effectively becomes a "maximum" sibling count. `siblingCount` of undefined would include as many sibling links as can fit.

### PR DESCRIPTION
**Pull request by jellison**
DRAFT: some unit tests are missing an implementation

**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [ ] For new widgets, an entry has been added to the `.dojorc`
* [ ] For new widgets, `theme.variant()` is added to the root domnode
* [ ] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [ ] WidgetProperties are exported

**Description:**

Pagination widget now responsive by automatically removing sibling links when space is insufficient to render `siblingCount` links.

When `siblingCount` is unset, all siblings are included which will fit the available space.

Resolves #1239 

##### Dojo theme
image

##### Material theme
image

### Mid-sized

image

### Min-sized

image

### CODE REVIEW

File: src/pagination/index.tsx [hunk 1]
```diff
@@ -12,17 +20,27 @@ import { ListOption } from '../list';
 import bundle from './Pagination.nls';
 import * as css from '../theme/default/pagination.m.css';
 
-function createMemoryTemplate<S = void>(): DataTemplate<S> {
+export function createMemoryTemplate<S = void>({
+	filter
+}: { filter?: (query: ResourceQuery[], v: S) => boolean } = {}): DataTemplate<S> {
 	return {
-		read: (_, put, get) => {
+		read: ({ query }, put, get) => {
```
Comment (tomdye) [inline src/pagination/index.tsx:27]: what made you add this in? I don't believe you're using the filter or query at all are you?

File: src/pagination/index.tsx [hunk 2]
```diff
@@ -53,20 +71,27 @@ interface PaginationCache {
 	currentPage: number;
 	pageSize: number;
 	pageSizes: number[];
-	pageSizesData: ListOption[];
+	pageSizesResource: { resource: () => Resource; data: ListOption[] };
```
Comment (tomdye) [inline src/pagination/index.tsx:74]: I think this should stay as `ListOption[]`, there's no need to expose the resource usage to the consumer of this pagination widget

File: src/pagination/index.tsx [hunk 3]
```diff
@@ -84,105 +109,167 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+			createMemoryResourceWithData(pageSizes.map((ps) => ({ value: ps.toString() })))
+		);
 	}
 
 	const page = icache.getOrSet('currentPage', 1);
 	const pageSize = icache.getOrSet('pageSize', 10);
-	const pageSizesData = icache.get('pageSizesData');
+	const pageSizesResource = icache.get('pageSizesResource');
+
+	const getRenderedWidth = (dnode: RenderResult) => {
```
Comment (tomdye) [inline src/pagination/index.tsx:122]: this should be outside of the factory so it's only created once

File: src/pagination/index.tsx [hunk 4]
```diff
@@ -84,105 +109,167 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+		</div>
+	);
 
-	const leading: WNode[] = [];
-	const leadingCount = Math.min(siblingCount, page - 1);
-	for (let i = 1; i <= leadingCount; i++) {
-		const linkTo = page - i;
+	let availableWidth =
+		containerWidth -
+		getRenderedWidth(prevLink) -
```
Comment (tomdye) [inline src/pagination/index.tsx:197]: all of these sizes should really be saved in an `icache` and not calculated every time, just on first render.

File: src/pagination/index.tsx [hunk 5]
```diff
@@ -84,105 +109,167 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+		getRenderedWidth(current) -
+		getRenderedWidth(nextLink);
 
-		leading.unshift(
+	for (let i = 1; i <= Math.max(leadingCount, trailingCount); i++) {
```
Comment (tomdye) [inline src/pagination/index.tsx:201]: Feels to me that this should just calculate what there's room for and render it, this trial and error approach will be costly. Should just calc and store the width of elements without buttons, and the width of a button then on resize, calculate how many fill fit, perhaps use a threshold of 90% available space? When calculating the width of the button up front you could use the largest page number as it's label, then you can always be sure they'll fit.

File: src/pagination/index.tsx (hunk 1 above)
Comment (jellison) [reply src/pagination/index.tsx:27]: No, I just copied it from examples.

File: src/pagination/index.tsx (hunk 3 above)
Comment (jellison) [reply src/pagination/index.tsx:122]: I had it outside, but it needs `theme.variant()` to render properly. I could just pass that in for each case.

File: src/pagination/index.tsx (hunk 2 above)
Comment (jellison) [reply src/pagination/index.tsx:74]: The intention was to only create the source when necessary (ie, when the `pageSizeData` had changed).

File: src/pagination/index.tsx (hunk 4 above)
Comment (jellison) [reply src/pagination/index.tsx:197]: I'll try it. I worry there will still be scenarios where the rendered width changes, such as changing themes in Parade (although I can't think of any others).

Comment (tomdye) [reply src/pagination/index.tsx:197]: you could clear the widths in the icache when the theme changes

File: src/pagination/index.tsx (hunk 2 above)
Comment (jellison) [reply src/pagination/index.tsx:74]: I think I've incorporated correct resource usage, but I'm still storing the resource in the cache. I'll let you resolve this conversation if you agree.

File: src/pagination/index.tsx [hunk 6]
```diff
@@ -51,22 +42,61 @@ interface PaginationCache {
...
 	currentPage: number;
+	nodeWidths: { prev: number; next: number; current: number; sibling: number };
 	pageSize: number;
 	pageSizes: number[];
-	pageSizesData: ListOption[];
+	pageSizesResource: { resource: () => Resource; data: ListOption[] };
+	theme?: Theme | ThemeWithVariant;
+}
+
+function createMemoryResourceWithData<S = any>(data: S[]) {
```
Comment (tomdye) [inline src/pagination/index.tsx:52]: This was used as a helper in the examples but realistically you should be creating a single resource outside of your render and then passing the data to it. You shouldn't be creating a whole new resource each time.

So here, just do

```ts
const template = {
				read: (_, put, get) => {
					let data: any[] = get();
					put(0, data);
					return { data, total: data.length };
				}
			};

const resource = createResource(template);
```

Then when you use it, pass it as
```
resource={{ resource: () => resource, data }}
```

The helper for the first part will be coming very soon via framework, today hopefully

File: src/pagination/index.tsx [hunk 7]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
 
 	if (pageSizes !== undefined && pageSizes !== icache.get('pageSizes')) {
 		icache.set('pageSizes', pageSizes);
-		icache.set('pageSizesData', pageSizes.map((ps) => ({ value: ps.toString() })));
+		icache.set(
+			'pageSizesResource',
+			createMemoryResourceWithData(pageSizes.map((ps) => ({ value: ps.toString() })))
```
Comment (tomdye) [inline src/pagination/index.tsx:119]: shouldn't be doing this here, see my comment above

File: src/pagination/index.tsx [hunk 8]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+			</div>
+			<div classes={classes.label}>{messages.next}</div>
+		</button>
+	);
+
+	const leadingSiblings: RenderResult[] = [];
+	const trailingSiblings: RenderResult[] = [];
+	const containerWidth = dimensions.get('links').size.width;
+	if (containerWidth) {
+		const nodeWidths = icache.getOrSet('nodeWidths', () => ({
```
Comment (tomdye) [inline src/pagination/index.tsx:174]: I think this would be better as individual icache entries.
Is it possible to just get the widget width when rendered with zero pages, and the width of a single page link and use those to do the maths? This seems like a fair few things to be checking. 
It also appears that it's duplicating the render code for the buttons / pages etc which could easily get out of sync.

File: src/pagination/index.tsx [hunk 9]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+
+		const maxLeading = siblingCount ? Math.min(siblingCount, page - 1) : page - 1;
+		const maxTrailing = siblingCount ? Math.min(siblingCount, total - page) : total - page;
+		let numSiblings = 0;
+
+		for (let i = 1; i <= Math.max(maxLeading, maxTrailing); i++) {
+			const showLeading = i <= maxLeading;
+			const showTrailing = i <= maxTrailing;
+
+			if (maxSiblings >= numSiblings + +showLeading + +showTrailing) {
```
Comment (tomdye) [inline src/pagination/index.tsx:202]: this is quite hard to follow

File: src/pagination/index.tsx [hunk 10]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+				if (showLeading) {
+					leadingSiblings.unshift(
+						<button
+							type="button"
+							key={`numberedLink-${page - i}`}
```
Comment (tomdye) [inline src/pagination/index.tsx:207]: perhaps create `previousPageNumber` and `nextPageNumber` variables?

File: src/pagination/index.tsx [hunk 11]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+							classes={[classes.numberedLink, classes.link]}
+							onclick={(e) => {
+								e.stopPropagation();
+								icache.set('currentPage', page - i);
+								onPage && onPage(page - i);
+							}}
+						>
+							{(page - i).toString()}
```
Comment (tomdye) [inline src/pagination/index.tsx:215]:
```
{`${previousPage}`}
```

File: src/pagination/index.tsx [hunk 12]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
+							type="button"
+							key={`numberedLink-${page + i}`}
+							classes={[classes.numberedLink, classes.link]}
+							onclick={(e) => {
+								e.stopPropagation();
+								icache.set('currentPage', page + i);
+								onPage && onPage(page + i);
+							}}
+						>
+							{(page + i).toString()}
```
Comment (tomdye) [inline src/pagination/index.tsx:233]:
```
{`${nextPage}`}
```

File: src/pagination/index.tsx (hunk 8 above)
Comment (jellison) [reply src/pagination/index.tsx:174]: The difficulty is that "zero page width" changes for the first and last pages because there are no previous and next links, respectively.

File: src/pagination/index.tsx [hunk 13]
```diff
@@ -84,105 +114,158 @@ export default factory(function Pagination({ middleware: { theme, icache, i18n }
...
 	if (pageSizes !== undefined && pageSizes !== icache.get('pageSizes')) {
 		icache.set('pageSizes', pageSizes);
-		icache.set('pageSizesData', pageSizes.map((ps) => ({ value: ps.toString() })));
+		icache.set(
+			'pageSizesResource',
+			createMemoryResourceWithData(pageSizes.map((ps) => ({ value: ps.toString() })))
+		);
+	}
+
+	if (theme.get() !== icache.get('theme')) {
```
Comment (tomdye) [inline src/pagination/index.tsx:123]: I think you should be able to use `diffProperty` here for this.
Could you try:

```ts
diffProperty(‘theme’, (current, next) => {
     if (current.theme !== next.theme) { 
      // recalculate here?
    }
});
```

File: src/pagination/index.tsx (hunk 8 above)
Comment (jellison) [reply src/pagination/index.tsx:174]: I'll clean up the whole loop and we can revisit.

File: src/pagination/index.tsx (hunk 6 above)
Comment (jellison) [reply src/pagination/index.tsx:52]: Got it, thanks!

File: src/pagination/index.tsx (hunk 13 above)
Comment (jellison) [reply src/pagination/index.tsx:123]: Good call. I should use that more.
